### PR TITLE
[Fix - ADF 142]: "Uncaught Error: no editor attached to the DOM element" in browser console after deleting Gap Match Interaction from canvas.

### PR DESCRIPTION
**Pull request by habemuscode**
**Issue**

https://oat-sa.atlassian.net/browse/ADF-142

**Description**

When the user removes Gap Match interaction with assigned response from the canvas, there is "Uncaught Error: no editor attached to the DOM element" error appears in the browser console.

**Steps to reproduce**

- Go to Items tab.
- Create a new item and go to it's Authoring.
- Add the Gap Match interaction to the canvas.
- Create a gap, assign a choice on it in 'Response' section and delete the Interaction.

**Actual result**

The interaction is deleted, the error message shows up in the browser console after some time.

**Expected result**

The interaction is deleted, no error messages in the console.

**Discussion**

2 solutions:

- Edit `htmlEditor` testing that other interactions are not affected. Other interactions are: `hottextInteraction` and `gapMatch` but in another part of the code

image

- Refactor in Question file and be sure we don't break `QTI` compatibility

```
 /**
* Workaround for ckEditor. When adding text to an empty choice, when initialising the widget on an empty content,
* ck automatically wraps it in a <p>, which breaks QTI compatibility.
* Therefore, we replace an empty choice with a placeholder text.
*/
if (! htmlEditor.getData($editable) || htmlEditor.getData($editable) === '') {
    placeholder = interaction.getNextPlaceholder();
    htmlEditor.setData($editable, placeholder);
}
```

**Bonus**

- Removed the error found in `mediaInteraction`, check comments in issue ticket

### CODE REVIEW

**Comment by habemuscode** [comment]
Deployed temporarily here: http://test-bosa.playground.kitchen.it.taocloud.org:49151 Ping me if not solved, maybe overwritten deployment

Solution 1 deployed

**Review by bziondik — CHANGES_REQUESTED** [review]
1. We don't commit bundles. They will be created on release
2. I think it is wrong place to remove throw error in `htmlEditor.js` , because it is used in lot of places. Can you check files related to  Gap Match Interaction and prevent calling functions `getData` and `setData` for this case

**Comment by habemuscode** [comment]
@bziondik 

- Bundles were pushed only to be able to test in Playground. I will remove it after getting approvals
- I will check your suggestion

Thanks

File: views/js/qtiCreator/editor/ckEditor/htmlEditor.js
Comment (jsconan) [inline views/js/qtiCreator/editor/ckEditor/htmlEditor.js:636]: **nitpick:** the function will always return `false`, no matter it performed the action or not.

File: views/js/qtiCreator/editor/ckEditor/htmlEditor.js
Comment (jsconan) [inline views/js/qtiCreator/editor/ckEditor/htmlEditor.js:621]: **nitpick:** Not sure this is the best approach. The return value will be either a boolean or a string... Why not keeping the exception and properly handle it on the consumer?

**Review by jsconan — CHANGES_REQUESTED** [review]
Quick code review only.

I wonder if this is the correct approach. Why not handle the execution properly on the consumer? Moreover, the API may be already used in several places, relying on the exception.

Then the change to the `setData()` function is not accurate as the state is not properly reflected.

File: views/js/qtiCreator/editor/ckEditor/htmlEditor.js
Comment (habemuscode) [reply views/js/qtiCreator/editor/ckEditor/htmlEditor.js]: ok

File: views/js/qtiCreator/editor/ckEditor/htmlEditor.js
Comment (habemuscode) [reply views/js/qtiCreator/editor/ckEditor/htmlEditor.js]: new solution will be provided

**Comment by habemuscode** [comment]
@bziondik @jsconan New solution pushed. I create another validation in the consumer that covers that case following your suggestions. Thanks

**Review by 404passionFruit — APPROVED** [review]

**Comment by 404passionFruit** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] The feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request's target is not `master`
- [x] Pull request title and description are meaningful
- [x] Commits are following conventional commits

**Comment by 404passionFruit** [comment]
Tested locally, issue is still there 
image

File: views/js/qtiCreator/widgets/choices/gapText/states/Question.js
Comment (404passionFruit) [inline views/js/qtiCreator/widgets/choices/gapText/states/Question.js:89]: Seems like this is not enough(

**Review by 404passionFruit — CHANGES_REQUESTED** [review]

**Review by bziondik — APPROVED** [review]
- [ ]  New code is covered by tests (if applicable)
- [x]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x]  New code is respecting code style rules
- [x]  New code is respecting best practices
- [x]  New code is not subject to concurrency issues (if applicable)
- [x]  Feature is working correctly on my local machine (if applicable)
- [x]  Acceptance criteria are respected
- [x]  Pull request title and description are meaningful

**Review by bziondik — CHANGES_REQUESTED** [review]
Failed, still errors
image

**Review by 404passionFruit — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] The feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request's target is not `master`
- [x] Pull request title and description are meaningful
- [x] Commits are following conventional commits

**Review by bziondik — APPROVED** [review]
Works!

**Pull request merged**
